File: server/routes/file-management.js
const express = require("express");
const multer = require("multer");
const { authenticateToken } = require("../middleware/auth");
const FileStorageService = require("../services/FileStorageService");
const { Configuration } = require("../models");
const ConfigurationService = require("../services/ConfigurationService");

const router = express.Router();
const fs = require('fs').promises;
const path = require('path');

// Configure multer for single file upload
const upload = multer({
  storage: multer.memoryStorage(),
  limits: {
    fileSize: Infinity, // No file size limit
  },
  fileFilter: (req, file, cb) => {
    // Accept all file types
    cb(null, true);
  }
});

// POST /api/file-management/upload - Upload a new file to a configuration
router.post("/upload", authenticateToken, upload.single('file'), async (req, res) => {
  try {
    const { configId, propertyPath } = req.body;
    const file = req.file;

    console.log(`File upload request: configId=${configId}, propertyPath=${propertyPath}`);

    if (!file) {
      return res.status(400).json({
        success: false,
        error: 'No file uploaded'
      });
    }

    if (!configId || !propertyPath) {
      return res.status(400).json({
        success: false,
        error: 'Missing configId or propertyPath'
      });
    }

    // Get the current configuration
    const config = await Configuration.findById(configId);
    if (!config) {
      return res.status(404).json({
        success: false,
        error: 'Configuration not found'
      });
    }

    const fileStorage = new FileStorageService();

    // Store the new file
    const fileMetadata = await fileStorage.storeFile(
      file.buffer,
      file.originalname,
      file.mimetype || 'application/octet-stream'
    );

    // Generate download URL
    const downloadUrl = await fileStorage.generateDownloadUrl(fileMetadata);

    // Create new file object
    const newFileObject = {
      _type: 'file',
      _metadata: {
        originalName: file.originalname,
        mimeType: file.mimetype || 'application/octet-stream',
        size: file.size,
        storageKey: fileMetadata.fileId,
        storageType: fileMetadata.storageType
      },
      _link: downloadUrl
    };

    // Update the configuration with the new file object
    // For nested paths, we need to build the proper nested structure
    const dataUpdate = {};
    setValueAtPath(dataUpdate, propertyPath, newFileObject);

    console.log(`Updating configuration ${configId} with file data:`, {
      propertyPath,
      dataUpdate: JSON.stringify(dataUpdate, null, 2),
      fileObject: JSON.stringify(newFileObject, null, 2)
    });

    await ConfigurationService.updateConfiguration(
      configId,
      {
        data: dataUpdate
      },
      "system" // Use system as updater for file uploads
    );

    console.log(`File upload completed: ${file.originalname} -> ${fileMetadata.fileId}`);

    res.json({
      success: true,
      message: `File "${file.originalname}" uploaded successfully`,
      metadata: newFileObject._metadata,
      downloadUrl: downloadUrl
    });

  } catch (error) {
    console.error("File upload error:", error);
    res.status(500).json({
      success: false,
      error: "Failed to upload file",
      details: error.message
    });
  }
});

// POST /api/file-management/replace - Replace a file in a configuration
router.post("/replace", authenticateToken, upload.single('file'), async (req, res) => {
  try {
    const { configId, propertyPath } = req.body;
    const file = req.file;

    console.log(`File replacement request: configId=${configId}, propertyPath=${propertyPath}`);

    if (!file) {
      return res.status(400).json({ 
        success: false,
        error: 'No file uploaded' 
      });
    }

    if (!configId || !propertyPath) {
      return res.status(400).json({ 
        success: false,
        error: 'Missing configId or propertyPath' 
      });
    }

    // Get the current configuration
    const config = await Configuration.findById(configId);
    if (!config) {
      return res.status(404).json({ 
        success: false,
        error: 'Configuration not found' 
      });
    }

    // Get the current value at the property path to find the old file
    const resolved = await ConfigurationService.resolveConfiguration(configId, false);
    const currentValue = getValueAtPath(resolved.resolved, propertyPath);

    // Check if current value is a file object
    if (!currentValue || currentValue._type !== 'file') {
      return res.status(400).json({ 
        success: false,
        error: 'Property is not a file object' 
      });
    }

    const fileStorage = new FileStorageService();

    // Delete the old file if it exists
    if (currentValue._metadata && currentValue._metadata.storageKey) {
      try {
        await fileStorage.deleteFile(currentValue._metadata);
        console.log('Deleted old file:', currentValue._metadata.storageKey);
      } catch (error) {
        console.warn('Failed to delete old file:', error.message);
        // Continue with replacement even if old file deletion fails
      }
    }

    // Store the new file
    const fileMetadata = await fileStorage.storeFile(
      file.buffer,
      file.originalname,
      file.mimetype || 'application/octet-stream'
    );

    // Generate download URL
    const downloadUrl = await fileStorage.generateDownloadUrl(fileMetadata);

    // Create new file object
    const newFileObject = {
      _type: 'file',
      _metadata: {
        originalName: file.originalname,
        mimeType: file.mimetype || 'application/octet-stream',
        size: file.size,
        storageKey: fileMetadata.fileId,
        storageType: fileMetadata.storageType
      },
      _link: downloadUrl
    };

    // Update the configuration with the new file object
    // For nested paths, we need to build the proper nested structure
    const dataUpdate = {};
    setValueAtPath(dataUpdate, propertyPath, newFileObject);

    await ConfigurationService.updateConfiguration(
      configId,
      {
        data: dataUpdate
      },
      "system" // Use system as updater for file replacements
    );

    console.log(`File replacement completed: ${file.originalname} -> ${fileMetadata.fileId}`);

    res.json({
      success: true,
      message: `File "${file.originalname}" uploaded successfully`,
      fileData: newFileObject
    });

  } catch (error) {
    console.error("File replacement error:", error);
    res.status(500).json({ 
      success: false,
      error: "Failed to replace file",
      details: error.message 
    });
  }
});

// Helper function to get value at a specific path
function getValueAtPath(obj, path) {
  if (!obj || !path) return obj;

  const keys = path.split('.');
  let current = obj;

  for (const key of keys) {
    if (current && typeof current === 'object' && current.hasOwnProperty(key)) {
      current = current[key];
    } else {
      return null;
    }
  }

  return current;
}

// Helper function to set value at a specific path
function setValueAtPath(obj, path, value) {
  if (!path) {
    return obj;
  }

  const keys = path.split('.');
  let current = obj;

  // Navigate to the parent of the target key
  for (let i = 0; i < keys.length - 1; i++) {
    const key = keys[i];
    if (!current[key] || typeof current[key] !== 'object') {
      current[key] = {};
    }
    current = current[key];
  }

  // Set the final value
  const finalKey = keys[keys.length - 1];
  current[finalKey] = value;

  return obj;
}

// GET /api/file-management/unreferenced - Find files not referenced in any configuration
router.get("/unreferenced", authenticateToken, async (req, res) => {
  try {
    console.log('Finding unreferenced files...');

    const fileStorage = new FileStorageService();

    // Get all stored files
    const allFiles = await getAllStoredFiles(fileStorage);
    console.log(`Found ${allFiles.length} total files in storage`);

    // Get all file references from configurations
    const referencedFiles = await getAllReferencedFiles();
    console.log(`Found ${referencedFiles.size} files referenced in configurations`);

    // Find unreferenced files
    const unreferencedFiles = allFiles.filter(file => !referencedFiles.has(file.storageKey));

    console.log(`Found ${unreferencedFiles.length} unreferenced files`);

    res.json({
      success: true,
      unreferencedFiles: unreferencedFiles.map(file => ({
        storageKey: file.storageKey,
        originalName: file.originalName,
        size: file.size,
        uploadDate: file.uploadDate,
        mimeType: file.mimeType
      })),
      totalFiles: allFiles.length,
      referencedFiles: referencedFiles.size,
      unreferencedCount: unreferencedFiles.length
    });

  } catch (error) {
    console.error('Failed to find unreferenced files:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to find unreferenced files'
    });
  }
});

// DELETE /api/file-management/unreferenced - Delete unreferenced files
router.delete("/unreferenced", authenticateToken, async (req, res) => {
  try {
    console.log('Deleting unreferenced files...');

    const fileStorage = new FileStorageService();

    // Get all stored files
    const allFiles = await getAllStoredFiles(fileStorage);
    console.log(`Found ${allFiles.length} total files in storage`);

    // Get all file references from configurations
    const referencedFiles = await getAllReferencedFiles();
    console.log(`Found ${referencedFiles.size} files referenced in configurations`);

    // Find unreferenced files
    const unreferencedFiles = allFiles.filter(file => !referencedFiles.has(file.storageKey));

    console.log(`Deleting ${unreferencedFiles.length} unreferenced files`);

    let deletedCount = 0;
    const errors = [];

    for (const file of unreferencedFiles) {
      try {
        await fileStorage.deleteFile({ storageKey: file.storageKey });
        deletedCount++;
        console.log(`Deleted unreferenced file: ${file.originalName} (${file.storageKey})`);
      } catch (error) {
        console.error(`Failed to delete file ${file.storageKey}:`, error);
        errors.push({
          storageKey: file.storageKey,
          originalName: file.originalName,
          error: error.message
        });
      }
    }

    res.json({
      success: true,
      message: `Successfully deleted ${deletedCount} unreferenced files`,
      deletedCount,
      totalCandidates: unreferencedFiles.length,
      errors
    });

  } catch (error) {
    console.error('Failed to delete unreferenced files:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to delete unreferenced files'
    });
  }
});

// Helper function to get all files from storage
async function getAllStoredFiles(fileStorage) {
  if (fileStorage.storageType === 'embedded') {
    const storageDir = fileStorage.embeddedStoragePath;

    try {
      const files = await fs.readdir(storageDir);
      const fileList = [];

      // Process metadata files
      for (const fileName of files) {
        if (fileName.endsWith('.meta.json')) {
          try {
            const metaFilePath = path.join(storageDir, fileName);
            const metaData = JSON.parse(await fs.readFile(metaFilePath, 'utf8'));
            const actualFileName = fileName.replace('.meta.json', '');
            const actualFilePath = path.join(storageDir, actualFileName);

            // Check if actual file exists
            try {
              await fs.access(actualFilePath);
              fileList.push({
                storageKey: metaData.storageKey || actualFileName,
                originalName: metaData.originalName,
                mimeType: metaData.mimeType,
                size: metaData.size,
                uploadDate: metaData.uploadDate,
                fileName: actualFileName
              });
            } catch (fileError) {
              console.warn(`File ${actualFileName} metadata exists but file missing`);
            }
          } catch (error) {
            console.warn(`Error reading metadata file ${fileName}:`, error.message);
          }
        }
      }

      return fileList;
    } catch (error) {
      console.warn('Error accessing storage directory:', error.message);
      return [];
    }
  } else {
    // TODO: Implement S3 file listing if needed
    console.warn('S3 unreferenced file cleanup not yet implemented');
    return [];
  }
}

// Helper function to get all file references from configurations
async function getAllReferencedFiles() {
  const referencedFiles = new Set();

  try {
    const configurations = await Configuration.findAll();

    for (const config of configurations) {
      if (config.data) {
        findFileReferences(config.data, referencedFiles);
      }
    }
  } catch (error) {
    console.error('Error getting configurations:', error);
  }

  return referencedFiles;
}

// Recursive function to find file references in configuration data
function findFileReferences(obj, referencedFiles) {
  if (typeof obj !== 'object' || obj === null) {
    return;
  }

  if (Array.isArray(obj)) {
    for (const item of obj) {
      findFileReferences(item, referencedFiles);
    }
  } else {
    // Check if this is a file object
    if (obj._type === 'file' && obj._metadata && obj._metadata.storageKey) {
      referencedFiles.add(obj._metadata.storageKey);
    }

    // Recursively check all properties
    for (const key in obj) {
      if (obj.hasOwnProperty(key)) {
        findFileReferences(obj[key], referencedFiles);
      }
    }
  }
}

module.exports = router;

File: server-go/internal/http/handlers/file_management.go
package handlers

import (
	"fmt"
	"io"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"

	"your.module/config-manager/internal/files"
	"your.module/config-manager/internal/logger"
	"your.module/config-manager/internal/mongo"
	"your.module/config-manager/internal/types"
)

type FileManagementHandler struct {
	storage *files.StorageManager
	db      *mongo.Client
	log     *logger.Logger
}

func NewFileManagementHandler(storage *files.StorageManager, db *mongo.Client, log *logger.Logger) *FileManagementHandler {
	return &FileManagementHandler{
		storage: storage,
		db:      db,
		log:     log,
	}
}

func (h *FileManagementHandler) Register(api *gin.RouterGroup) {
	fm := api.Group("/file-management")
	{
		fm.POST("/upload", h.upload)
		fm.POST("/replace", h.replace)
		fm.DELETE("/:storageKey", h.deleteFile)
		fm.GET("/unreferenced", h.listUnreferenced)
		fm.DELETE("/unreferenced", h.deleteUnreferenced)
	}
}

// POST /api/file-management/upload - Upload a new file to a configuration
func (h *FileManagementHandler) upload(c *gin.Context) {
	configID := c.PostForm("configId")
	propertyPath := c.PostForm("propertyPath")

	if configID == "" || propertyPath == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Missing configId or propertyPath",
		})
		return
	}

	// Get uploaded file
	file, err := c.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "No file uploaded",
		})
		return
	}

	// Verify configuration exists
	oid, err := primitive.ObjectIDFromHex(configID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Invalid configId",
		})
		return
	}

	var config types.Configuration
	err = h.db.Configurations.FindOne(c, bson.M{"_id": oid}).Decode(&config)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{
			"success": false,
			"error":   "Configuration not found",
		})
		return
	}

	// Open the file
	fileReader, err := file.Open()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to read uploaded file",
		})
		return
	}
	defer fileReader.Close()

	// Read file content
	fileContent, err := io.ReadAll(fileReader)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to read file content",
		})
		return
	}

	// Determine mime type
	mimeType := file.Header.Get("Content-Type")
	if mimeType == "" {
		mimeType = "application/octet-stream"
	}

	// Store the file
	fileMetadata, err := h.storage.StoreFile(c.Request.Context(), fileContent, file.Filename, mimeType)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to store file",
			"details": err.Error(),
		})
		return
	}

	// Generate download URL
	downloadURL, err := h.storage.GenerateDownloadURL(c.Request.Context(), fileMetadata)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to generate download URL",
		})
		return
	}

	// Create file object
	fileObject := map[string]interface{}{
		"_type": "file",
		"_metadata": map[string]interface{}{
			"originalName": file.Filename,
			"mimeType":     mimeType,
			"size":         file.Size,
			"storageKey":   fileMetadata.FileID,
			"storageType":  fileMetadata.StorageType,
		},
		"_link": downloadURL,
	}

	// Update configuration with file object at property path
	dataUpdate := make(map[string]interface{})
	setValueAtPath(dataUpdate, propertyPath, fileObject)

	// Perform update
	update := bson.M{
		"$set": bson.M{
			"data": dataUpdate,
		},
	}

	_, err = h.db.Configurations.UpdateOne(c, bson.M{"_id": oid}, update)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to update configuration",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success":     true,
		"message":     fmt.Sprintf("File \"%s\" uploaded successfully", file.Filename),
		"metadata":    fileObject["_metadata"],
		"downloadUrl": downloadURL,
	})
}

// POST /api/file-management/replace - Replace an existing file
func (h *FileManagementHandler) replace(c *gin.Context) {
	configID := c.PostForm("configId")
	propertyPath := c.PostForm("propertyPath")

	if configID == "" || propertyPath == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Missing configId or propertyPath",
		})
		return
	}

	// Get uploaded file
	file, err := c.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "No file uploaded",
		})
		return
	}

	// Verify configuration exists
	oid, err := primitive.ObjectIDFromHex(configID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Invalid configId",
		})
		return
	}

	var config types.Configuration
	err = h.db.Configurations.FindOne(c, bson.M{"_id": oid}).Decode(&config)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{
			"success": false,
			"error":   "Configuration not found",
		})
		return
	}

	// Get current value at property path to find old file
	currentValue := getValueAtPath(config.Data, propertyPath)
	if currentValue == nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Property path not found",
		})
		return
	}

	// Check if it's a file object
	currentMap, ok := currentValue.(map[string]interface{})
	if !ok || currentMap["_type"] != "file" {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Property is not a file object",
		})
		return
	}

	// Delete old file if it exists
	if metadata, ok := currentMap["_metadata"].(map[string]interface{}); ok {
		if storageKey, ok := metadata["storageKey"].(string); ok {
			fileMeta := &files.FileMetadata{StorageKey: storageKey}
			err := h.storage.DeleteFile(c.Request.Context(), fileMeta)
			if err != nil {
				fmt.Printf("Warning: Failed to delete old file %s: %v\n", storageKey, err)
				// Continue with replacement even if deletion fails
			}
		}
	}

	// Open the new file
	fileReader, err := file.Open()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to read uploaded file",
		})
		return
	}
	defer fileReader.Close()

	// Read file content
	fileContent, err := io.ReadAll(fileReader)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to read file content",
		})
		return
	}

	// Determine mime type
	mimeType := file.Header.Get("Content-Type")
	if mimeType == "" {
		mimeType = "application/octet-stream"
	}

	// Store the new file
	fileMetadata, err := h.storage.StoreFile(c.Request.Context(), fileContent, file.Filename, mimeType)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to store file",
			"details": err.Error(),
		})
		return
	}

	// Generate download URL
	downloadURL, err := h.storage.GenerateDownloadURL(c.Request.Context(), fileMetadata)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to generate download URL",
		})
		return
	}

	// Create new file object
	fileObject := map[string]interface{}{
		"_type": "file",
		"_metadata": map[string]interface{}{
			"originalName": file.Filename,
			"mimeType":     mimeType,
			"size":         file.Size,
			"storageKey":   fileMetadata.FileID,
			"storageType":  fileMetadata.StorageType,
		},
		"_link": downloadURL,
	}

	// Update configuration
	dataUpdate := make(map[string]interface{})
	setValueAtPath(dataUpdate, propertyPath, fileObject)

	update := bson.M{
		"$set": bson.M{
			"data": dataUpdate,
		},
	}

	_, err = h.db.Configurations.UpdateOne(c, bson.M{"_id": oid}, update)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to update configuration",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success":  true,
		"message":  fmt.Sprintf("File \"%s\" replaced successfully", file.Filename),
		"fileData": fileObject,
	})
}

// DELETE /api/file-management/:storageKey - Delete a file
func (h *FileManagementHandler) deleteFile(c *gin.Context) {
	storageKey := c.Param("storageKey")

	if storageKey == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Missing storageKey",
		})
		return
	}

	// Delete the file
	fileMeta := &files.FileMetadata{
		StorageKey: storageKey,
	}

	err := h.storage.DeleteFile(c.Request.Context(), fileMeta)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to delete file",
			"details": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"message": "File deleted successfully",
	})
}

// GET /api/file-management/unreferenced - List unreferenced files
func (h *FileManagementHandler) listUnreferenced(c *gin.Context) {
	// Get all stored files
	allFiles, err := h.storage.ListAllFiles(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to list files",
		})
		return
	}

	// Get all referenced files from configurations
	referencedFiles, err := h.getAllReferencedFiles(c)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to get referenced files",
		})
		return
	}

	// Find unreferenced files
	var unreferenced []map[string]interface{}
	for _, file := range allFiles {
		storageKey, _ := file["storageKey"].(string)
		if !referencedFiles[storageKey] {
			unreferenced = append(unreferenced, file)
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"success":           true,
		"unreferencedFiles": unreferenced,
		"totalFiles":        len(allFiles),
		"referencedFiles":   len(referencedFiles),
		"unreferencedCount": len(unreferenced),
	})
}

// DELETE /api/file-management/unreferenced - Delete all unreferenced files
func (h *FileManagementHandler) deleteUnreferenced(c *gin.Context) {
	// Get all stored files
	allFiles, err := h.storage.ListAllFiles(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to list files",
		})
		return
	}

	// Get all referenced files from configurations
	referencedFiles, err := h.getAllReferencedFiles(c)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to get referenced files",
		})
		return
	}

	// Delete unreferenced files
	deletedCount := 0
	var errors []map[string]interface{}

	for _, file := range allFiles {
		storageKey, _ := file["storageKey"].(string)
		if !referencedFiles[storageKey] {
			fileMeta := &files.FileMetadata{StorageKey: storageKey}
			err := h.storage.DeleteFile(c.Request.Context(), fileMeta)
			if err != nil {
				errors = append(errors, map[string]interface{}{
					"storageKey":   storageKey,
					"originalName": file["originalName"],
					"error":        err.Error(),
				})
			} else {
				deletedCount++
			}
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"success":         true,
		"message":         fmt.Sprintf("Successfully deleted %d unreferenced files", deletedCount),
		"deletedCount":    deletedCount,
		"totalCandidates": len(allFiles) - len(referencedFiles),
		"errors":          errors,
	})
}

// Helper: get all referenced files from configurations
func (h *FileManagementHandler) getAllReferencedFiles(c *gin.Context) (map[string]bool, error) {
	referencedFiles := make(map[string]bool)

	cursor, err := h.db.Configurations.Find(c, bson.M{})
	if err != nil {
		return nil, err
	}
	defer cursor.Close(c)

	for cursor.Next(c) {
		var config types.Configuration
		if err := cursor.Decode(&config); err != nil {
			continue
		}

		if config.Data != nil {
			findFileReferences(config.Data, referencedFiles)
		}
	}

	return referencedFiles, nil
}

// Helper: recursively find file references in data
func findFileReferences(data interface{}, referenced map[string]bool) {
	switch v := data.(type) {
	case map[string]interface{}:
		// Check if this is a file object
		if v["_type"] == "file" {
			if metadata, ok := v["_metadata"].(map[string]interface{}); ok {
				if storageKey, ok := metadata["storageKey"].(string); ok {
					referenced[storageKey] = true
				}
			}
		}
		// Recursively check all properties
		for _, value := range v {
			findFileReferences(value, referenced)
		}
	case []interface{}:
		for _, item := range v {
			findFileReferences(item, referenced)
		}
	}
}

// Helper: set value at a dot-separated path
func setValueAtPath(obj map[string]interface{}, path string, value interface{}) {
	keys := strings.Split(path, ".")
	current := obj

	// Navigate to parent, creating intermediate objects
	for i := 0; i < len(keys)-1; i++ {
		key := keys[i]
		if _, exists := current[key]; !exists {
			current[key] = make(map[string]interface{})
		}
		if next, ok := current[key].(map[string]interface{}); ok {
			current = next
		} else {
			// Overwrite non-map value
			current[key] = make(map[string]interface{})
			current = current[key].(map[string]interface{})
		}
	}

	// Set final value
	finalKey := keys[len(keys)-1]
	current[finalKey] = value
}

// Helper: get value at a dot-separated path
func getValueAtPath(obj map[string]interface{}, path string) interface{} {
	if obj == nil || path == "" {
		return obj
	}

	keys := strings.Split(path, ".")
	var current interface{} = obj

	for _, key := range keys {
		if m, ok := current.(map[string]interface{}); ok {
			current = m[key]
		} else {
			return nil
		}
	}

	return current
}

